Program 2: Data Wrangling and Transformation Using Pandas (Filtering, Grouping,
Aggregation, Sorting, and Reshaping)

1. Aim
To perform data wrangling and transformation operations using the Pandas library,
including filtering, grouping, aggregation, sorting, and reshaping of a dataset to prepare it for
effective analysis and decision-making.

2. Objectives
After completing this experiment, students will be able to:
• Understand the concept of data wrangling and transformation.
• Load and explore datasets using Pandas.
• Filter records based on specified conditions.
• Sort data in ascending and descending order.
• Group data based on one or more columns.
• Perform aggregation functions such as sum, mean, minimum, maximum, and count.
• Reshape datasets using pivot tables and melting techniques.
• Prepare structured datasets for analysis and machine learning.

3. Theory
3.1 Introduction to Data Wrangling
Data Wrangling (also known as Data Munging) is the process of converting raw, unstructured,
or inconsistent data into a clean and organized format suitable for analysis, visualization, and
machine learning.
In real-world applications, raw datasets often contain:
• Missing values
• Duplicate records
• Incorrect formats
• Unnecessary columns
• Inconsistent values
• Unorganized structures
Data wrangling improves the quality and usability of the data by performing various
transformation operations.

3.2 Importance of Data Wrangling
Data wrangling is essential because it:
• Improves data quality.
• Makes datasets easier to analyze.
• Enhances machine learning model performance.
• Removes inconsistencies and redundancy.
• Supports accurate business decision-making.
• Simplifies visualization and reporting.

3.3 Data Transformation
Data transformation refers to modifying the structure, format, or values of data to meet
analytical requirements.
Common transformation operations include:
• Filtering
• Sorting
• Grouping
• Aggregation
• Reshaping
• Renaming columns
• Creating new features

3.4 Filtering Data
Filtering is used to select only those records that satisfy a specified condition.
Example
Suppose a company wants to identify employees earning more than ₹45,000.

3.5 Sorting Data
Sorting arranges records in ascending or descending order.
Ascending Order
df.sort_values("Salary")
Descending Order
df.sort_values("Salary", ascending=False)
Sorting helps identify:
• Highest salary
• Lowest sales
• Top-performing products
• Latest transactions

3.6 Grouping Data
Grouping combines records having similar values.
Example:
City Salary
Delhi 40000
Delhi 45000
Mumbai 50000
Grouping by City:
Delhi
Mumbai
Syntax

df.groupby("City")

3.7 Aggregation
Aggregation summarizes grouped data.
Common aggregation functions include:
Function Description
sum() Total
mean() Average
max() Highest value
min() Lowest value
count() Number of records
median() Middle value
std() Standard deviation
Example:
df.groupby("City")["Salary"].mean()
Output:
City Average Salary
Delhi 42500
Mumbai 50000

3.8 Reshaping Data
Reshaping changes the structure of a dataset.
Common techniques include:
Pivot Table
Converts rows into columns.
pd.pivot_table(df,
values="Salary",
index="Department",
columns="City",
aggfunc="mean")

Melt
Converts columns into rows.
pd.melt(df)

4. Algorithm
- Import the Pandas library.
- Read the CSV dataset.
- Display the dataset.
- Filter records based on conditions.
- Sort the dataset.
- Group records by category.
- Perform aggregation functions.
- Create a pivot table.
- Perform melting.
- Display the transformed dataset.

5. Requirements
Software
• Python 3.x
• Jupyter Notebook / VS Code / Google Colab

Library
pip install pandas

Python Program

In [1]:
from google.colab import files
uploaded = files.upload()

Saving employees.csv to employees.csv


In [2]:
import pandas as pd

# Read dataset
df = pd.read_csv("employees.csv")

print("Original Dataset")
print(df)
# ------------------------------
# Filtering
# ------------------------------
print("\nEmployees with Salary > 45000")
filtered = df[df["Salary"] > 45000]
print(filtered)

# ------------------------------
# Sorting
# ------------------------------

print("\nSorted by Salary (Descending)")
sorted_df = df.sort_values(by="Salary", ascending=False)
print(sorted_df)
# ------------------------------
# Grouping
# ------------------------------
print("\nGrouped by Department")
grouped = df.groupby("Department")
# ------------------------------
# Aggregation
# ------------------------------
print("\nAverage Salary by Department")
print(grouped["Salary"].mean())
print("\nMaximum Salary by Department")
print(grouped["Salary"].max())
print("\nEmployee Count by Department")
print(grouped["Salary"].count())
# ------------------------------
# Pivot Table
# ------------------------------
print("\nPivot Table")
pivot = pd.pivot_table(
df,
values="Salary",
index="Department",
columns="City",
aggfunc="mean"
)
print(pivot)
# ------------------------------
# Melt
# ------------------------------

print("\nMelted Dataset")
melted = pd.melt(
df,
id_vars=["ID", "Name"],
value_vars=["Department", "City"]
)
print(melted)

Original Dataset
    ID   Name Department  Salary       City
0  101   John         HR   35000  Bangalore
1  102  Alice         IT   42000     Mumbai
2  103    Raj         IT   52000      Delhi
3  104  David    Finance   50000     Mumbai
4  105   Riya         HR   45000  Bangalore
5  106    Sam    Finance   60000      Delhi
6  107  Anita         IT   48000     Mumbai
7  108  Rohan         HR   39000      Delhi

Employees with Salary > 45000
    ID   Name Department  Salary    City
2  103    Raj         IT   52000   Delhi
3  104  David    Finance   50000  Mumbai
5  106    Sam    Finance   60000   Delhi
6  107  Anita         IT   48000  Mumbai

Sorted by Salary (Descending)
    ID   Name Department  Salary       City
5  106    Sam    Finance   60000      Delhi
2  103    Raj         IT   52000      Delhi
3  104  David    Finance   50000     Mumbai
6  107  Anita         IT   48000     Mumbai
4  105   Riya         HR   45000  Bangalore
1  102  Alice         IT   42000     Mumbai
7  108  Roha

Code Explanation
Import Pandas
import pandas as pd
Imports the Pandas library.

Read CSV
pd.read_csv()
Loads the dataset into a DataFrame.

Filtering
df[df["Salary"] > 45000]
Displays employees whose salary exceeds ₹45,000.

Sorting
sort_values() Arranges records in ascending or descending order.

Grouping
groupby() Groups employees according to their department.

Aggregation
mean()
max()
count()
Calculates average salary, maximum salary, and employee count.

Pivot Table
pivot_table() Summarizes data by department and city.

Melt
pd.melt() Converts multiple columns into a long-format dataset.

9. Sample Output
Filtered Data
ID Name Department Salary City
103 Raj IT 52000 Delhi
104 David Finance 50000 Mumbai
106 Sam Finance 60000 Delhi
107 Anita IT 48000 Mumbai

Average Salary by Department
Department
Finance 55000
HR 39667
IT 47333

Maximum Salary by Department
Department
Finance 60000
HR 45000
IT 52000

Employee Count
Department
Finance 2
HR 3
IT 3

Pivot Table
Department Bangalore Delhi Mumbai
Finance — 60000 50000
HR 40000 39000 —
IT — 52000 45000

Melted Data (Partial)
ID Name Variable Value
101 John Department HR
102 Alice Department IT
101 John City Bangalore
102 Alice City Mumbai

Result
The dataset was successfully transformed using Pandas by performing filtering, sorting,
grouping, aggregation, and reshaping operations. These transformations improved the
organization of the data and enabled meaningful analysis, making the dataset suitable for
reporting, visualization, and machine learning applications.